sdk-base/src/main/java/com/mapbox/maps/plugin/viewport/experimental/ViewportPlugin.kt

```diff
@@ -0,0 +1,180 @@
+package com.mapbox.maps.plugin.viewport.experimental
+
+import com.mapbox.maps.plugin.MapPlugin
+import com.mapbox.maps.plugin.viewport.DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS
+import com.mapbox.maps.plugin.viewport.DEFAULT_STATE_TRANSITION_MAX_DURATION_MS
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportDataSource
+import com.mapbox.maps.plugin.viewport.experimental.state.FollowingViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.OverviewViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportState
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportStateChangedObserver
+import com.mapbox.maps.plugin.viewport.experimental.transition.TransitionEndListener
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransitionOptions
+
+/**
+ * The Viewport plugin allows to track objects on a map.
+ *
+ * Most common use-case is the ability to track the user location and animate the location updates
+ * to create a navigation experience. Next to tracking a single object, the API allows tracking
+ * multiple objects. The viewport needs to automatically optimize the camera position to take in
+ * account the user provided data.
+ */
+interface ViewportPlugin : MapPlugin {
+
+  /**
+   * Returns current [ViewportState].
+   *
+   * If current state is null, returns null.
+   * @see registerViewportStateChangedObserver
+   */
+  val currentState: ViewportState?
+
+  /**
+   * Returns list of registered states.
+   */
+  val states: Set<ViewportState>
+
+  /**
+   * Add a viewport state to the viewport plugin, which could be reused later with the state id.
+   *
+   * @param state the view port state to be added.
+   */
+  fun addViewportState(state: ViewportState)
+
+  /**
+   * Remove a viewport state from the viewport plugin, and it could not be reused later with the state id.
+
+   * @param state the view port state to be removed.
+   */
+  fun removeViewportState(state: ViewportState)
```

Comment:
If we would make state instance a key, we wouldn't need to register instances before usage.

Reply:
Removed these APIs in https://github.com/mapbox/mapbox-maps-android/pull/1034